Improve performance of StringText.Write & StringBuilderText.Write on .NET Core

Neme12 opened the pull request:
In `StringText`, the change is to call the span overload in case of a partial text span as opposed to delegating to the base method that uses `CopyTo`. In `StringBuilderText`, the method is new and uses the overload that takes a `StringBuilder` in case of a full text span, otherwise enumerates and writes the individual chunks, again avoiding the copying.

There is one concern though and it's that if the `TextWriter` is a custom type that doesn't override the `ReadOnlySpan<char>` and `StringBuilder` overloads, this might actually be slower. Is that a common use case or do we expect this to be mostly used with the built-in `TextWriter` classes that override everything appropriately?

Neme12 commented:
Here are benchmarks for `StringText.Write`. First, I copied the method, named it `WriteNew` and only changed the implementation of the new one, so the methods could be benchmarked at the same time.

// Writing to StringWriter

```
|          Method |                  Job |              Runtime |      Mean |    Error |   StdDev | Ratio | RatioSD |
|---------------- |--------------------- |--------------------- |----------:|---------:|---------:|------:|--------:|
|       WriteFull |             .NET 6.0 |             .NET 6.0 |  69.43 us | 1.380 us | 1.589 us |  1.00 |    0.00 |
|    WriteFullNew |             .NET 6.0 |             .NET 6.0 |  69.67 us | 1.368 us | 1.463 us |  1.00 |    0.03 |
|    WritePartial |             .NET 6.0 |             .NET 6.0 | 121.42 us | 1.280 us | 1.198 us |  1.75 |    0.05 |
| WritePartialNew |             .NET 6.0 |             .NET 6.0 |  68.93 us | 1.046 us | 0.979 us |  1.00 |    0.03 |
|                 |                      |                      |           |          |          |       |         |
|       WriteFull |        .NET Core 3.1 |        .NET Core 3.1 |  72.23 us | 1.290 us | 1.207 us |  1.00 |    0.00 |
|    WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 |  72.07 us | 0.521 us | 0.435 us |  1.00 |    0.02 |
|    WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 110.34 us | 1.211 us | 1.074 us |  1.53 |    0.03 |
| WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 |  72.83 us | 1.024 us | 0.958 us |  1.01 |    0.02 |
|                 |                      |                      |           |          |          |       |         |
|       WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 |  70.82 us | 1.380 us | 1.643 us |  1.00 |    0.00 |
|    WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 |  70.34 us | 1.065 us | 0.996 us |  0.99 |    0.03 |
|    WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 111.57 us | 1.239 us | 1.159 us |  1.57 |    0.04 |
| WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 111.81 us | 1.776 us | 1.661 us |  1.57 |    0.05 |
```

```c#
BenchmarkRunner.Run<SourceTextBenchmark>();

[SimpleJob(RuntimeMoniker.Net472)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net60)]
public class SourceTextBenchmark
{
    private readonly SourceText _sourceText;
    private readonly TextSpan _fullSpan;
    private readonly TextSpan _partialSpan;

    private readonly StringBuilder _stringBuilder;
    private readonly StringWriter _stringWriter;

    public SourceTextBenchmark()
    {
        _sourceText = SourceText.From(File.ReadAllText(@"C:\Users\{username}\source\repos\dotnet\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs"));

        _fullSpan = TextSpan.FromBounds(0, _sourceText.Length);
        _partialSpan = TextSpan.FromBounds(4, _sourceText.Length);

        _stringBuilder = new StringBuilder();
        _stringWriter = new StringWriter(_stringBuilder);
    }

    [Benchmark(Baseline = true)]
    public void WriteFull()
    {
        _stringBuilder.Clear();
        _sourceText.Write(_stringWriter, _fullSpan);
    }

    [Benchmark]
    public void WriteFullNew()
    {
        _stringBuilder.Clear();
        _sourceText.WriteNew(_stringWriter, _fullSpan);
    }

    [Benchmark]
    public void WritePartial()
    {
        _stringBuilder.Clear();
        _sourceText.Write(_stringWriter, _partialSpan);
    }

    [Benchmark]
    public void WritePartialNew()
    {
        _stringBuilder.Clear();
        _sourceText.WriteNew(_stringWriter, _partialSpan);
    }
}
```

// Writing to a file using StreamWriter(FileStream)

```
|          Method |                  Job |              Runtime |     Mean |     Error |    StdDev | Ratio | RatioSD |
|---------------- |--------------------- |--------------------- |---------:|----------:|----------:|------:|--------:|
|       WriteFull |             .NET 6.0 |             .NET 6.0 | 1.045 ms | 0.0162 ms | 0.0152 ms |  1.00 |    0.00 |
|    WriteFullNew |             .NET 6.0 |             .NET 6.0 | 1.028 ms | 0.0141 ms | 0.0125 ms |  0.98 |    0.02 |
|    WritePartial |             .NET 6.0 |             .NET 6.0 | 1.093 ms | 0.0144 ms | 0.0135 ms |  1.05 |    0.02 |
| WritePartialNew |             .NET 6.0 |             .NET 6.0 | 1.031 ms | 0.0120 ms | 0.0112 ms |  0.99 |    0.02 |
|                 |                      |                      |          |           |           |       |         |
|       WriteFull |        .NET Core 3.1 |        .NET Core 3.1 | 1.069 ms | 0.0170 ms | 0.0159 ms |  1.00 |    0.00 |
|    WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 | 1.052 ms | 0.0129 ms | 0.0121 ms |  0.98 |    0.02 |
|    WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 1.141 ms | 0.0100 ms | 0.0094 ms |  1.07 |    0.02 |
| WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 | 1.054 ms | 0.0124 ms | 0.0116 ms |  0.99 |    0.02 |
|                 |                      |                      |          |           |           |       |         |
|       WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.543 ms | 0.0183 ms | 0.0171 ms |  1.00 |    0.00 |
|    WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.536 ms | 0.0194 ms | 0.0182 ms |  1.00 |    0.02 |
|    WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.589 ms | 0.0188 ms | 0.0175 ms |  1.03 |    0.02 |
| WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.586 ms | 0.0159 ms | 0.0149 ms |  1.03 |    0.02 |
```

```c#
BenchmarkRunner.Run<SourceTextBenchmark>();

[SimpleJob(RuntimeMoniker.Net472)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net60)]
public class SourceTextBenchmark
{
    private readonly SourceText _sourceText;
    private readonly TextSpan _fullSpan;
    private readonly TextSpan _partialSpan;

    private readonly FileStream _fileStream;
    private readonly StreamWriter _streamWriter;

    public SourceTextBenchmark()
    {
        _sourceText = SourceText.From(File.ReadAllText(@"C:\Users\{username}\source\repos\dotnet\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs"));

        _fullSpan = TextSpan.FromBounds(0, _sourceText.Length);
        _partialSpan = TextSpan.FromBounds(4, _sourceText.Length);

        var fileName = Path.GetTempFileName();
        _fileStream = File.OpenWrite(fileName);
        _streamWriter = new StreamWriter(_fileStream);
    }

    [Benchmark(Baseline = true)]
    public void WriteFull()
    {
        _fileStream.Position = 0;
        _sourceText.Write(_streamWriter, _fullSpan);
    }

    [Benchmark]
    public void WriteFullNew()
    {
        _fileStream.Position = 0;
        _sourceText.WriteNew(_streamWriter, _fullSpan);
    }

    [Benchmark]
    public void WritePartial()
    {
        _fileStream.Position = 0;
        _sourceText.Write(_streamWriter, _partialSpan);
    }

    [Benchmark]
    public void WritePartialNew()
    {
        _fileStream.Position = 0;
        _sourceText.WriteNew(_streamWriter, _partialSpan);
    }
}
```

Neme12 commented:
Here are benchmarks for `StringBuilderText.Write`:

// Writing to StringWriter

```
|          Method |                  Job |              Runtime |      Mean |    Error |   StdDev | Ratio | RatioSD |
|---------------- |--------------------- |--------------------- |----------:|---------:|---------:|------:|--------:|
|       WriteFull |             .NET 6.0 |             .NET 6.0 | 131.94 us | 0.881 us | 0.824 us |  1.00 |    0.00 |
|    WriteFullNew |             .NET 6.0 |             .NET 6.0 |  73.23 us | 0.717 us | 0.599 us |  0.56 |    0.01 |
|    WritePartial |             .NET 6.0 |             .NET 6.0 | 126.25 us | 1.224 us | 1.145 us |  0.96 |    0.01 |
| WritePartialNew |             .NET 6.0 |             .NET 6.0 |  77.85 us | 0.642 us | 0.569 us |  0.59 |    0.01 |
|                 |                      |                      |           |          |          |       |         |
|       WriteFull |        .NET Core 3.1 |        .NET Core 3.1 | 127.70 us | 1.736 us | 1.624 us |  1.00 |    0.00 |
|    WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 |  72.00 us | 0.956 us | 0.894 us |  0.56 |    0.01 |
|    WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 133.41 us | 1.386 us | 1.296 us |  1.04 |    0.02 |
| WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 |  75.84 us | 0.739 us | 0.691 us |  0.59 |    0.01 |
|                 |                      |                      |           |          |          |       |         |
|       WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 128.60 us | 0.479 us | 0.425 us |  1.00 |    0.00 |
|    WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 129.20 us | 0.738 us | 0.654 us |  1.00 |    0.01 |
|    WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 128.48 us | 0.814 us | 0.636 us |  1.00 |    0.01 |
| WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 129.23 us | 0.985 us | 0.921 us |  1.00 |    0.01 |
```

```c#
BenchmarkRunner.Run<SourceTextBenchmark>();

[SimpleJob(RuntimeMoniker.Net472)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net60)]
public class SourceTextBenchmark
{
    private readonly SourceText _sourceText;
    private readonly TextSpan _fullSpan;
    private readonly TextSpan _partialSpan;

    private readonly StringBuilder _stringBuilder;
    private readonly StringWriter _stringWriter;

    public SourceTextBenchmark()
    {
        SyntaxTree syntaxTree;

        using (var fileStream = File.OpenRead(@"C:\Users\{username}\source\repos\dotnet\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs"))
            syntaxTree = CSharpSyntaxTree.ParseText(SourceText.From(fileStream));

        _sourceText = syntaxTree.GetRoot().GetText();

        _fullSpan = TextSpan.FromBounds(0, _sourceText.Length);
        _partialSpan = TextSpan.FromBounds(4, _sourceText.Length);

        _stringBuilder = new StringBuilder();
        _stringWriter = new StringWriter(_stringBuilder);
    }

    [Benchmark(Baseline = true)]
    public void WriteFull()
    {
        _stringBuilder.Clear();
        _sourceText.Write(_stringWriter, _fullSpan);
    }

    [Benchmark]
    public void WriteFullNew()
    {
        _stringBuilder.Clear();
        _sourceText.WriteNew(_stringWriter, _fullSpan);
    }

    [Benchmark]
    public void WritePartial()
    {
        _stringBuilder.Clear();
        _sourceText.Write(_stringWriter, _partialSpan);
    }

    [Benchmark]
    public void WritePartialNew()
    {
        _stringBuilder.Clear();
        _sourceText.WriteNew(_stringWriter, _partialSpan);
    }
}
```

// Writing to a file using StreamWriter(FileStream)

```
|          Method |                  Job |              Runtime |     Mean |     Error |    StdDev | Ratio | RatioSD |
|---------------- |--------------------- |--------------------- |---------:|----------:|----------:|------:|--------:|
|       WriteFull |             .NET 6.0 |             .NET 6.0 | 1.133 ms | 0.0093 ms | 0.0087 ms |  1.00 |    0.00 |
|    WriteFullNew |             .NET 6.0 |             .NET 6.0 | 1.025 ms | 0.0090 ms | 0.0084 ms |  0.90 |    0.01 |
|    WritePartial |             .NET 6.0 |             .NET 6.0 | 1.139 ms | 0.0190 ms | 0.0178 ms |  1.01 |    0.02 |
| WritePartialNew |             .NET 6.0 |             .NET 6.0 | 1.029 ms | 0.0053 ms | 0.0049 ms |  0.91 |    0.01 |
|                 |                      |                      |          |           |           |       |         |
|       WriteFull |        .NET Core 3.1 |        .NET Core 3.1 | 1.147 ms | 0.0111 ms | 0.0098 ms |  1.00 |    0.00 |
|    WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 | 1.089 ms | 0.0014 ms | 0.0011 ms |  0.95 |    0.01 |
|    WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 1.149 ms | 0.0081 ms | 0.0072 ms |  1.00 |    0.01 |
| WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 | 1.067 ms | 0.0153 ms | 0.0136 ms |  0.93 |    0.02 |
|                 |                      |                      |          |           |           |       |         |
|       WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.601 ms | 0.0160 ms | 0.0134 ms |  1.00 |    0.00 |
|    WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.628 ms | 0.0219 ms | 0.0205 ms |  1.02 |    0.01 |
|    WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.607 ms | 0.0149 ms | 0.0139 ms |  1.01 |    0.01 |
| WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 1.592 ms | 0.0172 ms | 0.0161 ms |  1.00 |    0.01 |
```

```c#
BenchmarkRunner.Run<SourceTextBenchmark>();

[SimpleJob(RuntimeMoniker.Net472)]
[SimpleJob(RuntimeMoniker.NetCoreApp31)]
[SimpleJob(RuntimeMoniker.Net60)]
public class SourceTextBenchmark
{
    private readonly SourceText _sourceText;
    private readonly TextSpan _fullSpan;
    private readonly TextSpan _partialSpan;

    private readonly FileStream _fileStream;
    private readonly StreamWriter _streamWriter;

    public SourceTextBenchmark()
    {
        SyntaxTree syntaxTree;

        using (var fileStream = File.OpenRead(@"C:\Users\{username}\source\repos\dotnet\roslyn\src\Compilers\CSharp\Portable\Generated\BoundNodes.xml.Generated.cs"))
            syntaxTree = CSharpSyntaxTree.ParseText(SourceText.From(fileStream));

        _sourceText = syntaxTree.GetRoot().GetText();

        _fullSpan = TextSpan.FromBounds(0, _sourceText.Length);
        _partialSpan = TextSpan.FromBounds(4, _sourceText.Length);

        var fileName = Path.GetTempFileName();
        _fileStream = File.OpenWrite(fileName);
        _streamWriter = new StreamWriter(_fileStream);
    }

    [Benchmark(Baseline = true)]
    public void WriteFull()
    {
        _fileStream.Position = 0;
        _sourceText.Write(_streamWriter, _fullSpan);
    }

    [Benchmark]
    public void WriteFullNew()
    {
        _fileStream.Position = 0;
        _sourceText.WriteNew(_streamWriter, _fullSpan);
    }

    [Benchmark]
    public void WritePartial()
    {
        _fileStream.Position = 0;
        _sourceText.Write(_streamWriter, _partialSpan);
    }

    [Benchmark]
    public void WritePartialNew()
    {
        _fileStream.Position = 0;
        _sourceText.WriteNew(_streamWriter, _partialSpan);
    }
}
```

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,45 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+            else
+            {
+                if (span.Length == 0)
+                    return;
```
move this check to the top?

Neme12 replied:
Done.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,45 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+            if (span.Start == 0 && span.Length == this.Length)
+            {
+                textWriter.Write(_builder);
+            }
```
nit, consider doing an early return after the trivial case, that way it's clear nothing follows it.  it also enables the complex case to not need indentation.

Neme12 replied:
Done.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,45 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+                if (span.Length == 0)
+                    return;
+
+                int chunkOffset = 0;
```
doc waht this is.

Neme12 replied:
@CyrusNajmabadi  would you be ok with `// Index of the current chunk in the text.`?

Neme12 replied:
Or do you mean document the entire loop?

CyrusNajmabadi replied:
basically, i'm not sure what chunkOffset means.  offset is a very vague term that could be interpreted in a lot of ways.

Neme12 replied:
I didn't name it index because that would sound like it's the index an offset array.

Neme12 replied:
@jaredpar Do you have any suggestions for a better name?

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,45 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+
+                foreach (var chunk in _builder.GetChunks())
+                {
+                    var startFromChunk = span.Start - chunkOffset;
```
not sure how to read the term "startFromChunk".  sounds like a boolean from teh name.  can you clarify what this means?

Neme12 replied:
It's the starting index of the span, with respect to the current chunk (as opposed to the start of the text). How would you name this?

CyrusNajmabadi replied:
`var startingIndexInCurrentChunk`?

Neme12 replied:
@CyrusNajmabadi even though it might not actually be inside that chunk?

Neme12 replied:
I mean you could say that span.Start sounds like a boolean or an action. The variable is defined right away and it's just one minus, if you don't get it at the first sight, you'd be able to see what it is from the initialization? It's kinda weird that the name would be almost twice as long as the expression it's supposed to shorten.

CyrusNajmabadi replied:
i'm still not really sure what's going on here.  it doesn't help that this expression is built from another variable i'm unclear about "offset".

Neme12 replied:
@jaredpar Do you have any suggestions for a better name?

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,45 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+                        var startIndex = Math.Max(startFromChunk, 0);
+                        var endIndex = Math.Min(endFromChunk, chunk.Length);
+
+                        textWriter.Write(chunk.Span.Slice(startIndex, endIndex - startIndex));
```
```suggestion
                        textWriter.Write(chunk.Span.Slice(startIndex, endIndex - startIndex));
```

is this different from just `chunk.Span[startIndex..endIndex]`?

Neme12 replied:
No, I forgot this syntax exists, thanks.

Neme12 replied:
Done.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringText.cs:
```diff
@@ -92,11 +92,17 @@ public override void Write(TextWriter textWriter, TextSpan span, CancellationTok
-                textWriter.Write(this.Source);
+                // Even on .NET Core, prefer the string overload, because in case the writer is a custom TextWriter that doesn't
+                // override the ReadOnlySpan<char> overload, it delegates to the array one under the hood.
+                textWriter.Write(_source);
```
```suggestion
                textWriter.Write(this.Source);
```

Neme12 replied:
Why?

CyrusNajmabadi replied:
because the change seems entirely stylistic with no impact on behavior (and as such isn't necessary for this change around improving performance).

Neme12 replied:
Done.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Text/StringText.cs:
```diff
@@ -92,11 +92,17 @@ public override void Write(TextWriter textWriter, TextSpan span, CancellationTok
             if (span.Start == 0 && span.End == this.Length)
             {
-                textWriter.Write(this.Source);
+                // Even on .NET Core, prefer the string overload, because in case the writer is a custom TextWriter that doesn't
```
can you clarify what you mean by "the string overload"?

CyrusNajmabadi replied:
n/m.

jaredpar commented on src/Compilers/Core/CodeAnalysisTest/Text/StringBuilderTextTest.cs:
```diff
@@ -0,0 +1,52 @@
+            // across their boundaries.
+            stringBuilder.Append("chunk1");
+            stringBuilder.Append("chunk2");
+            stringBuilder.Append("chunk3");
```
Could we add an `Assert` here to verify that this produces at least two chunks in the data? The internal structure of chunks seems subject to change so would like some verification we test this code path over the long term.

Neme12 replied:
Done.

Neme12 commented on src/Compilers/Core/Portable/Text/StringBuilderText.cs:
```diff
@@ -89,5 +91,47 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+            CheckSubSpan(span);
+
+            if (span.Length == 0)
+                return;
```
These checks need to be repeated in each override, which isn't ideal. I thought about just having these in the base Write method and having it forward to a new WriteCore method that is actually virtual and overriden, but:
1. Write is a public method so making it non-virtual would be an API change. I guess this class can only be derived internally though so it could be OK?
2. With that change, if you have a derived class and call Write on it, it would now go through virtual dispatch, whereas now, it doesn't have to if the class is sealed.

jaredpar replied:
> I guess this class can only be derived internally though so it could be OK?

It can be really difficult to tell this by looking at the code. The rules for what we consider truly public vs. not are easy to mess up. The easiest way to tell though is look at `PublicAPI.Shipped.txt`. If the API is there then it's public and 3rd parties can derive from it. In this case it is public. 

https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PublicAPI.Shipped.txt#L3957

Neme12 replied:
@jaredpar It is public, but SourceText has internal virtual methods, which means that external code cannot derive from it, right?

jaredpar replied:
This is not easy to tell at a glance. I spent a good 5 minutes searching around before I wrote my previous reply because I couldn't remember if `SourceText` was derivable or not.

CyrusNajmabadi replied:
is it just this first test?  if so, it seems fine to repeat.  the code will be extremely clear and hard to mess up.

Neme12 replied:
@CyrusNajmabadi Yes, it's clear like this, but it's easy to forget to do this when someone overrides the method - eg. in `StringText`, the bounds checking was missing before this PR.

Neme12 commented:
Failure seems to be a flaky test. How can I rerun CI?

Neme12 commented:
/azp run

Neme12 commented:
> There is one concern though and it's that if the TextWriter is a custom type that doesn't override the ReadOnlySpan<char> and StringBuilder overloads, this might actually be slower. Is that a common use case or do we expect this to be mostly used with the built-in TextWriter classes that override everything appropriately?

To mitigate this, I could check whether the method is overriden and only then use the new behavior, but that would require reflection, and I'm not sure if that's supported on all platforms Roslyn targets and whether that would be too slow anyway.

Neme12 commented:
> There is one concern though and it's that if the TextWriter is a custom type that doesn't override the ReadOnlySpan<char> and StringBuilder overloads, this might actually be slower. Is that a common use case or do we expect this to be mostly used with the built-in TextWriter classes that override everything appropriately?

I addressed this concern. I benchmarked the new implementations on a custom `TextWriter` that wraps a `StringWriter` and overrides the string and array overloads, but not the span or string builder ones.

<details>
<summary>Here's the type.</summary>

```c#
sealed class CustomTextWriter : TextWriter
{
    private readonly TextWriter _textWriter;

    public CustomTextWriter(TextWriter textWriter)
    {
        _textWriter = textWriter;
    }

    public override Encoding Encoding =>
        _textWriter.Encoding;

    public override void Write(char value) =>
        _textWriter.Write(value);

    public override void Write(string? value) =>
        _textWriter.Write(value);

    public override void Write(char[] buffer, int index, int count) =>
        _textWriter.Write(buffer, index, count);

    public override void Write(char[]? buffer) =>
        _textWriter.Write(buffer);
}
```
</details>

Then I simply added new benchmarks that write to this wrapper type:
```c#
[Benchmark]
public void WriteFullNewCustomWriter()
{
    _stringBuilder.Clear();
    _sourceText.WriteNew(_customWriter, _fullSpan);
}
```

These were the results on .NET 6 - all of them writing to `StringWriter`.
 for `StringText.Write` (no Full versions there because the implementation for full spans stayed the same)
```
|                      Method |      Mean |    Error |   StdDev |
|---------------------------- |----------:|---------:|---------:|
|                WritePartial | 139.72 us | 2.759 us | 2.581 us |
|             WritePartialNew |  81.61 us | 1.346 us | 1.259 us |
| WritePartialNewCustomWriter | 215.75 us | 4.075 us | 4.185 us |
```
and for `StringBuilderText.Write`:
```
|                      Method |      Mean |    Error |   StdDev |    Median | Ratio | RatioSD |
|---------------------------- |----------:|---------:|---------:|----------:|------:|--------:|
|                   WriteFull | 133.33 us | 0.895 us | 0.794 us | 133.66 us |  1.00 |    0.00 |
|                WriteFullNew |  87.61 us | 1.751 us | 3.457 us |  88.40 us |  0.64 |    0.04 |
|    WriteFullNewCustomWriter | 140.58 us | 2.702 us | 2.395 us | 140.74 us |  1.05 |    0.02 |
|                WritePartial | 146.43 us | 1.766 us | 1.566 us | 146.42 us |  1.10 |    0.01 |
|             WritePartialNew |  83.90 us | 1.557 us | 3.760 us |  82.61 us |  0.67 |    0.02 |
| WritePartialNewCustomWriter | 124.85 us | 1.666 us | 1.558 us | 125.47 us |  0.94 |    0.01 |
```

It seems that with `StringText.Write`, when using a custom writer, it's about 2x slower compared to the original version. With `StringBuilderText.Write`, there doesn't seem to be a significant difference.

I changed the implementation of `StringText.Write` to fallback to the old implementation in case the writer doesn't override the span overload. Then, I benchmarked everything again.

`StringText.Write`:
```
|                      Method |                  Job |              Runtime |      Mean |    Error |   StdDev | Ratio | RatioSD |
|---------------------------- |--------------------- |--------------------- |----------:|---------:|---------:|------:|--------:|
|                   WriteFull |             .NET 6.0 |             .NET 6.0 |  83.25 us | 1.088 us | 1.017 us |  1.00 |    0.00 |
|                WriteFullNew |             .NET 6.0 |             .NET 6.0 |  83.37 us | 1.601 us | 1.844 us |  1.00 |    0.03 |
|    WriteFullNewCustomWriter |             .NET 6.0 |             .NET 6.0 |  69.73 us | 1.365 us | 1.822 us |  0.84 |    0.02 |
|                WritePartial |             .NET 6.0 |             .NET 6.0 | 119.52 us | 1.588 us | 1.485 us |  1.44 |    0.02 |
|             WritePartialNew |             .NET 6.0 |             .NET 6.0 |  71.67 us | 1.394 us | 1.304 us |  0.86 |    0.02 |
| WritePartialNewCustomWriter |             .NET 6.0 |             .NET 6.0 | 121.97 us | 1.390 us | 1.300 us |  1.47 |    0.02 |
|                             |                      |                      |           |          |          |       |         |
|                   WriteFull |        .NET Core 3.1 |        .NET Core 3.1 |  72.90 us | 0.829 us | 0.775 us |  1.00 |    0.00 |
|                WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 |  71.66 us | 1.393 us | 1.303 us |  0.98 |    0.02 |
|    WriteFullNewCustomWriter |        .NET Core 3.1 |        .NET Core 3.1 |  72.09 us | 1.205 us | 1.128 us |  0.99 |    0.02 |
|                WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 117.96 us | 1.634 us | 1.364 us |  1.62 |    0.03 |
|             WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 |  74.78 us | 1.248 us | 1.167 us |  1.03 |    0.02 |
| WritePartialNewCustomWriter |        .NET Core 3.1 |        .NET Core 3.1 | 116.97 us | 0.996 us | 0.931 us |  1.60 |    0.02 |
|                             |                      |                      |           |          |          |       |         |
|                   WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 |  71.43 us | 1.036 us | 0.969 us |  1.00 |    0.00 |
|                WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 |  71.55 us | 1.169 us | 1.093 us |  1.00 |    0.02 |
|    WriteFullNewCustomWriter | .NET Framework 4.7.2 | .NET Framework 4.7.2 |  71.99 us | 1.419 us | 1.518 us |  1.01 |    0.03 |
|                WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 119.72 us | 1.689 us | 1.410 us |  1.68 |    0.04 |
|             WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 121.66 us | 1.648 us | 1.542 us |  1.70 |    0.04 |
| WritePartialNewCustomWriter | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 120.46 us | 1.731 us | 1.535 us |  1.69 |    0.03 |
```
`StringBuilderText.Write`:
```
|                      Method |                  Job |              Runtime |      Mean |    Error |   StdDev | Ratio | RatioSD |
|---------------------------- |--------------------- |--------------------- |----------:|---------:|---------:|------:|--------:|
|                   WriteFull |             .NET 6.0 |             .NET 6.0 | 163.95 us | 3.259 us | 6.804 us |  1.00 |    0.00 |
|                WriteFullNew |             .NET 6.0 |             .NET 6.0 |  90.40 us | 1.735 us | 2.256 us |  0.54 |    0.03 |
|    WriteFullNewCustomWriter |             .NET 6.0 |             .NET 6.0 | 131.38 us | 1.660 us | 1.631 us |  0.77 |    0.02 |
|                WritePartial |             .NET 6.0 |             .NET 6.0 | 143.53 us | 1.365 us | 1.210 us |  0.84 |    0.03 |
|             WritePartialNew |             .NET 6.0 |             .NET 6.0 |  83.41 us | 0.879 us | 0.823 us |  0.49 |    0.02 |
| WritePartialNewCustomWriter |             .NET 6.0 |             .NET 6.0 | 129.16 us | 1.009 us | 0.895 us |  0.75 |    0.03 |
|                             |                      |                      |           |          |          |       |         |
|                   WriteFull |        .NET Core 3.1 |        .NET Core 3.1 | 141.26 us | 0.464 us | 0.434 us |  1.00 |    0.00 |
|                WriteFullNew |        .NET Core 3.1 |        .NET Core 3.1 |  77.88 us | 1.136 us | 1.063 us |  0.55 |    0.01 |
|    WriteFullNewCustomWriter |        .NET Core 3.1 |        .NET Core 3.1 | 131.29 us | 0.968 us | 0.906 us |  0.93 |    0.01 |
|                WritePartial |        .NET Core 3.1 |        .NET Core 3.1 | 140.53 us | 1.432 us | 1.269 us |  0.99 |    0.01 |
|             WritePartialNew |        .NET Core 3.1 |        .NET Core 3.1 |  81.54 us | 1.067 us | 0.891 us |  0.58 |    0.01 |
| WritePartialNewCustomWriter |        .NET Core 3.1 |        .NET Core 3.1 | 125.84 us | 1.368 us | 1.213 us |  0.89 |    0.01 |
|                             |                      |                      |           |          |          |       |         |
|                   WriteFull | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 138.48 us | 1.754 us | 1.464 us |  1.00 |    0.00 |
|                WriteFullNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 138.36 us | 0.867 us | 0.811 us |  1.00 |    0.01 |
|    WriteFullNewCustomWriter | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 138.15 us | 1.780 us | 1.665 us |  1.00 |    0.01 |
|                WritePartial | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 137.98 us | 1.707 us | 1.597 us |  1.00 |    0.02 |
|             WritePartialNew | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 138.88 us | 2.708 us | 2.533 us |  1.00 |    0.02 |
| WritePartialNewCustomWriter | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 137.03 us | 1.603 us | 1.499 us |  0.99 |    0.01 |
```

@CyrusNajmabadi @jaredpar

Neme12 commented on src/Compilers/Core/Portable/Text/StringText.cs:
```diff
@@ -90,14 +90,35 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
+#endif
         }
+
+        private delegate void TextWriterWriteReadOnlySpanDelegate(ReadOnlySpan<char> buffer);
```
It's a bit of a shame that I can't somehow specify the overload using parameter types when referring to a method. If I could, the "method group" could have a natural delegate type and I wouldn't have to declare one.

Neme12 commented:
> I really don't like the idea of baking in knowledge like this. Feels very hacky to me

I mean yes, it is a bit hacky, but that the reality of
a) backwords compat (if TextWriter was new, derived types would be forced to override the span overload - the base implementation is kind of a hack for back compat)
b) wanting more perf

CyrusNajmabadi commented:
> b) wanting more perf

The presumption is that this takes precedence over other concerns (like clarity/maintainability).  For example, if things get so complex that it's more likely to introduce bugs, the perf savings may not be worth it.  Perf is just one axis here.

Neme12 commented:
I don't think checking whether a method is overriden is really that complex. Also, `SourceText` is a core type that is that everything else is built upon, so I think being a bit more focus on performance in this particular part of the project is worth it.

CyrusNajmabadi commented:
> I don't think checking whether a method is overriden is really that complex. 

It def is.  AFAICT, we don't follow this pattern anywhere else in roslyn.   It also feels like it's definitely violating the idea of these types and inheritance in general.  I feel like this is very fragile.

Neme12 commented:
> I feel like this is very fragile

I don't understand. Why is this fragile? Implementers of `TextWriter` who know about spans will get the better behavior, others won't. It's not like the base implementation will ever become efficient because it has to delegate to the array one for back compat. We're essentially making sure to not call the back compat-only method.

CyrusNajmabadi commented:
Because it's tying behavior in one type to specific introspection about inheritance in another. I don't believe it's a good idea for one type to try to be inferring and things in this manner (especially by checking what types delegates point to). 

I don't believe we do this anywhere else in Roslyn and I'm skeptical that the perf here is worth it given the complexity. 

Put another way, if we take this, I would presume we would have to take similar PRs for all other subsystems, and I don't see a reason for us to necessarily want to do that. Like I mentioned before, perf is one aspect of code to have to consider. Maintainability and complexity are others.

Neme12 commented:
> given the complexity

I could see your argument about complexity if this was using things like `GetMethod` etc, but that's not the case. It's a quick simple check. Regarding your argument about fragility, could you present a potential future scenario that you're worried about that would break this?

> we would have to take similar PRs for all other subsystems

I think things like this would make sense in core performance sensitive parts like `SourceText`, the lexer, `SlidingTextWindow` etc.. As I said, SourceText is the basis of virtually everything that deals with source code in Roslyn so improvements here could have broad impact.

Neme12 commented:
> Because it's tying behavior in one type to specific introspection about inheritance in another. I don't believe it's a good idea for one type to try to be inferring and things in this manner

It seems like a reasonable assumption to me. `Write(ReadOnlySpan<char>)` is a new method that was added to an existing type that has to support inheritance. It has to call into overloads that take an array to maintain the behavior of the derived class, and the only way you can convert a span into an array is by allocating/pooling and copying the data, which is inefficient. If a derived class overrides the method, it's likely reading the data from the span directly rather than copying it to an array. Why would someone override it if not to make it more efficient. Seems like a totally reasonable assumption to me.

CyrusNajmabadi commented:
> It's a quick simple check. 

I don't find it simple at all. As mentioned, I don't believe we do this elsewhere. And, conceptually, having one type interrogate they inheritance hierarchy of some value passed to it is complex on the face of it as it now tightly couples the perf on one area to an implementation detail of another value and how it relates inheritance-wise to others.

CyrusNajmabadi commented:
> the lexer, SlidingTextWindow etc

Here's the thing, I don't believe those are core perf areas. For example, in the last 5+ years of diving and fixing real world perf issues, I don't believe we've seen and cases where the perf of these types was an issue. 

So implementing these types of changes for those types would be increasing complexity in areas that aren't benefiting real world scenarios. That comes with very significant implications then for maintainability. Specifically, these types are *already* complex. To the point that it makes implementing language features harder (I say this having had to deal with this heavily while implementing raw string literals).  So the perf benefit wouldn't benefit real world cases, and it would come with the true drawback of making it harder to ship language features. These aspects of any change have to be considered as we are not trying to ship perf as the primary deliverable of the code. Instead we are trying to ship a codebase that balances many concerns, where perf is just one thing.

CyrusNajmabadi commented:
> could have broad impact.

For perf, it would be better to not be speculative.  If we have real world scenarios that we can see demonstrative perf gains on, that would be helpful. 

For sourcetext I do believe that is possible as o believe we have seen some reports with source generators where there were issues. However it would be good to actually take those cases and see if these changes address the problem. Then the perf benefit for those cases can be appropriately compared against the high complexity and maintainability cost here.

CyrusNajmabadi commented:
> could you present a potential future scenario that you're worried about that would break this?

The cost is on maintainability. Anyone who has to maintain this now needs to hold all this complexity in their head and get things right on framework/core for any changes here in the future. That added complexity is more chance to get things wrong and to introduce subtle issues (like potential corruption).

Neme12 commented:
> it now tightly couples the perf on one area to an implementation detail of another value and how it relates inheritance-wise to others

But perf *always* depends on implementation details of any external code and libraries you're using, because those implementation details affect the perf. In the next version of a library, the function you're using might be 2x slower (although realistically, that would usually be treated as a regression in corefx and it's unlikely to happen in the first place because they take perf of every area really seriously and have automated tests to detect any regressions). But still, when you're comparing 2 ways of doing something and you find that 1 of them is faster, in the next version, it could be the other one. So with perf, you're *always* relying on implementation details. That's why I tested this with multiple .NET versions.

Most importantly though, I'm not relying on implementation details with respect to correctness of behavior, only with respect to desirable perf outcomes.

> The cost is on maintainability. Anyone who has to maintain this now needs to hold all this complexity in their head and get things right on framework/core for any changes here in the future. That added complexity is more chance to get things wrong and to introduce subtle issues (like potential corruption).

I really don't see where the complexity or maintainability cost is. You're talking about this piece of code:
```c#
var @delegate = (TextWriterWriteReadOnlySpanDelegate)textWriter.Write;
if (@delegate.Method.DeclaringType != typeof(TextWriter))
{
    // Use the Write(ReadOnlySpan<char>) overload only if the method is overriden, because if it's not, the base
    // TextWriter implementation needs to delegate to the array overload and could potentially allocate a really
    // large array, which would be slower than the buffering base implementation.
    textWriter.Write(this.Source.AsSpan(span.Start, span.Length));
}
else
{
    base.Write(textWriter, span, cancellationToken);
}
```
That's literally just 1 condition - to check whether the `TextWriter` *supports* writing directly from a span, and 2 branches based on that, each do the same thing and both 1 line (or rather the second one calls an existing methods). I really don't see where the cost is on maintainability. Both branches have the same behavior so if someone edits this someday and gets the condition wrong, it's not like the code will start behaving incorrectly or like there would be any corruption, as you're saying. It would just lose the perf benefit - that's literally it. The behavior would still be correct (and the tests also cover the behavior in both branches). And on top of that, this isn't the kind of method that has to evolve for new language features etc. This just copies a piece of data. This is the kind of code that is written once and usually stays the same because it doesn't need to evolve or change.

I'm kinda surprised by this. In dotnet/aspnetcore and especially dotnet/runtime, they routinely add all sorts of and hard to read convoluted code with many branches and different implementations to achieve perf benefits in any public API. This is really *nothing* compared to that. This is just one condition and 2 equally simple branches. I just don't see the complexity. I guess the recent perf enthusiasm in .NET hasn't reached Roslyn as much 😔

CyrusNajmabadi commented:
> I really don't see where the complexity or maintainability cost is

I've been through this code, and the apis (and impls) for sourcetext and textwriterAnd I'm still not convinced this pr is correct. That's a couple of hours so far and a ton more uncertainty and difficulty for the future for any changes in this area. That is literal cost already so far, and that's just on reviewing

Neme12 commented:
There are tests that verify that it's correct. Where do you think there could be an issue?

Neme12 commented:
@CyrusNajmabadi Would you be willing to take this without the check for the span overload being overriden? All the built-in .NET `TextWriter` classes implement it and the majority of uses should be with the built-in ones, and in case someone has a custom `TextWriter`, they should really provide that overload anyway.

I checked all non-test usages of `Write` in Roslyn, and 9 out of 11 write to `StreamWriter`. The rest write to `SourceTextWriter`, which can be fixed to implement it.

CyrusNajmabadi commented:
> @CyrusNajmabadi Would you be willing to take this 

I think the primary thing affecting if i think we should take this would be real world examples of how this benefits areas that are important perf hotspots.  If, for example, in generators, this would make a substantive difference on existing generators, that would be great.  However, if it would just be a barely noticeable difference that accounts to just noise wrt to where perf is being spent today, then i don't see the complexity as being worth it.

Can you collect any data in this regard?  Thanks!

CyrusNajmabadi commented:
Moving ot draft for now. @Neme12 If you'd like to pick this back up again, let us know!
